### PR TITLE
Fix requests with large bodies - fixes #220

### LINKED ISSUE #220: Large resources unusable with 2.0.0
We have an application that includes creating or applying a wide range or resources and resource types. After updating to 2.0.0 (and the tls changes you suggested, thanks!)  most things are working However we can't apply or put on large resources. This happens most often with CRD's but I'm sure I can get config maps that hit this size limit too.

Assume that we essentially do something like this:

```
{:ok, conn} = Conn.from_file(path)
conn = struct!(conn, insecure_skip_tls_verify: true)
create_operation = Client.create(resource)
Client.run(connection, create_operation)
```

That results in:

```elixir
    %K8s.Client.HTTPError{
      message:
        "the given data exceeds the request window size, which is 262144. The server will refill the window size of the request when ready. This will be handled transparently by stream/2.",
      adapter_specific_error: %Mint.HTTPError{
        reason: {:exceeds_window_size, :request, 262_144},
        module: Mint.HTTP2
      }
    }
 ```
 
 An example of a rather large resource CRD is:
 https://github.com/spotahome/redis-operator/blob/master/manifests/databases.spotahome.com_redisfailovers.yaml

### PR DESCRIPTION
**Pull request by mruoss**
As raised in issue #220 we currently have a problem with large resources / request bodies. 

- [x] Entry in CHANGELOG.md was created
- [ ] New unit tests were created
- [x] New integration tests were created

### CODE REVIEW

**Comment by mruoss** [comment]
@elliottneilclark FYI.

As mentioned in the issue #220, I've started by adding a failing test.

**Comment by mruoss** [comment]
I can't figure out how to make dialyzer happy with this code (see https://github.com/elixir-mint/mint/issues/380). Besides that, I think this should do the trick. @elliottneilclark would you care to test your code using this branch?

**Comment by mruoss** [comment]
Dialyzer is happy now. Ready to review.

File: lib/k8s/client/mint/dialyzer_hacks.ex [hunk 1]
```diff
@@ -0,0 +1,11 @@
+defmodule K8s.Client.Mint.DialyzerHacks do
+  @moduledoc """
+  Just some hacks to make the Dialyzer happy
+  """
+
+  @doc """
+  This function tells Dialyzer that the conn is actually a HTTP2 conn.
```
Comment (elliottneilclark) [inline lib/k8s/client/mint/dialyzer_hacks.ex]: Might want the why or the story of how this came to be, those questions are important but often lost to time.

File: lib/k8s/client/mint/request.ex [hunk 2]
```diff
@@ -155,4 +174,46 @@ defmodule K8s.Client.Mint.Request do
...
+      case Mint.HTTP.protocol(conn) do
+        :http1 ->
+          {pending_request_body, []}
+
+        :http2 ->
+          chunk_size = chunk_size(request, conn)
+          Enum.split(pending_request_body, chunk_size)
+      end
+
+    with {:ok, conn} <- Mint.HTTP.stream_request_body(conn, request_ref, chunk),
```
Comment (elliottneilclark) [inline lib/k8s/client/mint/request.ex:190]: Chunk here is not always the expected types of iodata. https://github.com/elixir-mint/mint/blob/v1.4.2/lib/mint/http2.ex#L508

When starting a watch on just about every kind on a cluster I'm seeing:
```
  * 1st argument: not an iodata term                                                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                                                               
            :erlang.iolist_size([123, 34, 97, 112, 105, 86, 101, 114, 115, 105, 111, 110, 34, 58, 34, 97, 112, 105, 101, 120, 116, 101, 110, 115, 105, 111, 110, 115, 46, 107, 56, 115, 46, 105, 111, 47, 118, 49, 34, 44, 34, 107, 105, 110, 100, 34, 58, 34, 67, 117, ...])                                  
            (mint 1.4.2) lib/mint/http2.ex:1196: Mint.HTTP2.encode_data/4                                                                                                                                                                                                                                      
            (mint 1.4.2) lib/mint/http2.ex:526: Mint.HTTP2.stream_request_body/3                                                                                                                                                                                                                               
            (k8s 2.0.0) lib/k8s/client/mint/request.ex:196: K8s.Client.Mint.Request.stream_request_body/2                                                                                                                                                                                                      
            (k8s 2.0.0) lib/k8s/client/mint/http_adapter.ex:171: K8s.Client.Mint.HTTPAdapter.handle_call/3                                                                                                                                                                                                     
            (stdlib 4.2) gen_server.erl:1149: :gen_server.try_handle_call/4                                                                                                                                                                                                                                    
            (stdlib 4.2) gen_server.erl:1178: :gen_server.handle_msg/6                                                                                                                                                                                                                                         
            (stdlib 4.2) proc_lib.erl:240: :proc_lib.init_p_do_apply/3                                                                                                                                                                                                                                         
    (elixir 1.14.2) lib/gen_server.ex:1038: GenServer.call/3                                                                                                                                                                                                                                                   
    (k8s 2.0.0) lib/k8s/client/mint_http_provider.ex:41: anonymous fn/3 in K8s.Client.MintHTTPProvider.stream/5                                                                                                                                                                                                
    (elixir 1.14.2) lib/stream.ex:1623: Stream.do_resource/5                                                                                                                                                                                                                                                   
    (elixir 1.14.2) lib/stream.ex:942: Stream.do_transform/5                                                                                                                                                                                                                                                   
    (elixir 1.14.2) lib/stream.ex:1811: Enumerable.Stream.do_each/4                                                                                                                                                                                                                                            
    (elixir 1.14.2) lib/stream.ex:942: Stream.do_transform/5                                                                                                                                                                                                                                                   
    (elixir 1.14.2) lib/enum.ex:4307: Enum.each/2                   
```

File: lib/k8s/client/mint/http_adapter.ex [hunk 3]
```diff
@@ -148,20 +148,30 @@ defmodule K8s.Client.Mint.HTTPAdapter do
...
-            Request.new(pool: pool, stream_to: stream_to, caller_ref: caller_ref)
-          )
-
-        {:reply, {:ok, request_ref}, struct!(state, conn: conn)}
-
+    # stream body if there is one.
+    {body, pending_request_body} =
+      case body do
+        nil -> {body, body}
+        body -> {:stream, String.to_charlist(body)}
```
Comment (elliottneilclark) [inline lib/k8s/client/mint/http_adapter.ex]: We don't need to stream if the entire request fits in the window. I'm pretty sure that saves a round trip, so might be a nice optimization to put as a TODO.

**Comment by elliottneilclark** [comment]
This looks super close, thank you! I'm seeing an issue after starting a watch and streaming the events.

File: lib/k8s/client/mint/request.ex (hunk 2 above)
Comment (mruoss) [reply lib/k8s/client/mint/request.ex:190]: Oh boy yeah that's me being ignorant about iodata / binary / string / charlist.

File: lib/k8s/client/mint/http_adapter.ex (hunk 3 above)
Comment (mruoss) [reply lib/k8s/client/mint/http_adapter.ex]: Indeed, nor if it's HTTP1. 
The HTTP1 case I'll do now. The other one probably a todo for now.

File: lib/k8s/client/mint/dialyzer_hacks.ex (hunk 1 above)
Comment (mruoss) [reply lib/k8s/client/mint/dialyzer_hacks.ex]: Indeed, thanks.

File: lib/k8s/client/mint/http_adapter.ex (hunk 3 above)
Comment (mruoss) [reply lib/k8s/client/mint/http_adapter.ex]: actually, not I can add both checks - non need for the TODO

**Comment by mruoss** [comment]
> This looks super close, thank you! I'm seeing an issue after starting a watch and streaming the events.

Oh boy yeah that's me being ignorant about iodata / binary / string / charlist. Added an emoticon to the annotation => test fails. But fixed it now. Together with the other feedback. Thanks for the review!

**Comment by mruoss** [comment]
Opened https://github.com/elixir-mint/mint/issues/381

**Review by elliottneilclark — APPROVED** [review]
Thank you!

**Pull request merged**